packages/block-editor/src/components/rich-text/index.native.js

```diff
@@ -445,6 +480,29 @@ function RichTextWrapper(
 				isURL( plainText.trim() )
 			) {
 				mode = 'BLOCKS';
+
+				createEmbed = () => {
```

Comment:
@fluiddot I haven't gotten the functionality here working in totality. The challenge I am having is when the bottom sheet is shown and the actions are clicked the code in these functions is not being run. I think the issue lies in how they are being referenced by the options that were declared above. At a glance, do you have any thoughts on this? 
I am wondering if I needed to utilize the spread operator to update the `onSelect` value within the `options` object. I am not sure. 

Once I get it working I will thoroughly test and leave other thoughts/comments for a final review.

Reply:
Yep, this case is a bit tricky because you need to update the picker options (specifically the `onSelect` property of each option) right before the picker is presented. At a first glance, I think this would require to probably keep the picker options in the local state so it re-renders the Picker component with the new options when a URL is pasted.

Reply:
After a second thought, since the `onSelect` callbacks are called in `onPickerSelect` handler ([reference](https://github.com/WordPress/gutenberg/pull/35204/files#diff-1bf4290b2ff8a9f2d347d3450576c98613f26f26f4ce05b850cf232483f7b28cR245)), we could use `useRef` to hold the picker options and modify the `onSelect` of each option when this code is called.

Reply:
I've been playing around with the approach I mentioned in the previous comment and I think we could also isolate the logic in a sub-component. Here is a patch diff with the approach I'm thinking, feel free to take it but keep in mind that I only did a couple of tests so it might require further improvements.

```patch
diff --git a/packages/block-editor/src/components/rich-text/index.native.js b/packages/block-editor/src/components/rich-text/index.native.js
index e9469f605137701f6a60446f48a5dea4a265266f..26e68eb338b0c1dbba9aa088efbb7a26c65241f5 100644
--- a/packages/block-editor/src/components/rich-text/index.native.js
+++ b/packages/block-editor/src/components/rich-text/index.native.js
@@ -14,7 +14,6 @@ import {
 	useCallback,
 	forwardRef,
 } from '@wordpress/element';
-import { Picker } from '@wordpress/components';
 import { useDispatch, useSelect } from '@wordpress/data';
 import {
 	pasteHandler,
@@ -40,7 +39,6 @@ import {
 } from '@wordpress/rich-text';
 import deprecated from '@wordpress/deprecated';
 import { isURL } from '@wordpress/url';
-import { __ } from '@wordpress/i18n';
 
 /**
  * Internal dependencies
@@ -59,6 +57,7 @@ import {
 	isShortcode,
 	createLinkInParagraph,
 } from './utils';
+import EmbedHandlerPicker from './embed-handler-picker';
 
 const wrapperClasses = 'block-editor-rich-text';
 const classes = 'block-editor-rich-text__editable';
@@ -121,6 +120,7 @@ function RichTextWrapper(
 	const fallbackRef = useRef();
 	const { clientId, isSelected: blockIsSelected } = useBlockEditContext();
 	const nativeProps = useNativeProps();
+	const embedHandlerPickerRef = useRef();
 	const selector = ( select ) => {
 		const {
 			isCaretWithinFormattedText,
@@ -213,38 +213,6 @@ function RichTextWrapper(
 			);
 	}
 
-	const embeddableUrlPickerRef = useRef();
-
-	let createEmbed = () => {
-	};
-	let createLink = () => {
-	};
-
-	let embeddableUrlPickerOptions = {
-		createEmbed: {
-			id: 'createEmbed',
-			label: __( 'Create embed' ),
-			value: 'createEmbed',
-			onSelect: createEmbed,
-		},
-		createLink: {
-			id: 'createLink',
-			label: __( 'Create link' ),
-			value: 'createLink',
-			onSelect: createLink,
-		},
-	};
-
-	const options = [
-		embeddableUrlPickerOptions.createEmbed,
-		embeddableUrlPickerOptions.createLink,
-	];
-
-	function onPickerSelect( value ) {
-		const selectedItem = options.find( ( item ) => item.value === value );
-		selectedItem.onSelect();
-	}
-
 	const onSelectionChange = useCallback(
 		( start, end ) => {
 			selectionChange( clientId, identifier, start, end );
@@ -480,28 +448,26 @@ function RichTextWrapper(
 				isURL( plainText.trim() )
 			) {
 				mode = 'BLOCKS';
-
-				createEmbed = () => {
-					const content = pasteHandler( {
-						HTML: html,
-						plainText,
-						mode,
-						tagName,
-						preserveWhiteSpace,
-					} );
-
-					if ( content.length > 0 ) {
-						if ( onReplace && isEmpty( value ) ) {
-							onReplace( content, content.length - 1, -1 );
+				// Embed handler
+				embedHandlerPickerRef.current?.presentPicker( {
+					createEmbed: () => {
+						const content = pasteHandler( {
+							HTML: html,
+							plainText,
+							mode,
+							tagName,
+							preserveWhiteSpace,
+						} );
+
+						if ( content.length > 0 ) {
+							if ( onReplace && isEmpty( value ) ) {
+								onReplace( content, content.length - 1, -1 );
+							}
 						}
-					}
-				};
-
-				createLink = () => {
-					createLinkInParagraph( plainText.trim(), onReplace );
-				};
-
-				embeddableUrlPickerRef.current?.presentPicker();
+					},
+					createLink: () =>
+						createLinkInParagraph( plainText.trim(), onReplace ),
+				} );
 				return;
 			}
 
@@ -708,13 +674,7 @@ function RichTextWrapper(
 							/>
 						) }
 					</Autocomplete>
-					<Picker
-						ref={ embeddableUrlPickerRef }
-						options={ options }
-						onChange={ onPickerSelect }
-						hideCancelButton
-						leftAlign
-					/>
+					<EmbedHandlerPicker ref={ embedHandlerPickerRef } />
 				</>
 			) }
 		</RichText>
diff --git a/packages/block-editor/src/components/rich-text/embed-handler-picker.native.js b/packages/block-editor/src/components/rich-text/embed-handler-picker.native.js
new file mode 100644
index 0000000000000000000000000000000000000000..07a77c7ba8f482a891feadf72cd5a1348e90118b
--- /dev/null
+++ b/packages/block-editor/src/components/rich-text/embed-handler-picker.native.js
@@ -0,0 +1,56 @@
+/**
+ * External dependencies
+ */
+import { noop } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { forwardRef, useRef, useImperativeHandle } from '@wordpress/element';
+import { Picker } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+
+const DEFAULT_PICKER_OPTIONS = [
+	{
+		id: 'createEmbed',
+		label: __( 'Create embed' ),
+		value: 'createEmbed',
+		onSelect: noop,
+	},
+	{
+		id: 'createLink',
+		label: __( 'Create link' ),
+		value: 'createLink',
+		onSelect: noop,
+	},
+];
+
+export default forwardRef( ( {}, ref ) => {
+	const pickerRef = useRef();
+	const pickerOptions = useRef( DEFAULT_PICKER_OPTIONS ).current;
+
+	function onPickerSelect( value ) {
+		const selectedItem = pickerOptions.find(
+			( item ) => item.value === value
+		);
+		selectedItem.onSelect();
+	}
+
+	useImperativeHandle( ref, () => ( {
+		presentPicker: ( { createEmbed, createLink } ) => {
+			pickerOptions[ 0 ].onSelect = createEmbed;
+			pickerOptions[ 1 ].onSelect = createLink;
+			pickerRef.current?.presentPicker();
+		},
+	} ) );
+
+	return (
+		<Picker
+			ref={ pickerRef }
+			options={ pickerOptions }
+			onChange={ onPickerSelect }
+			hideCancelButton
+			leftAlign
+		/>
+	);
+} );

```

Reply:
Amazing contribution here @fluiddot I knew the utilization of refs and forward ref would be able to isolate the `Picker` in a self-contained component and allow the passing of the options in a dynamic way but I just wasn't sure how to do it. I have learned an immense amount about how to utilize `useImperativeHandle` and `forwardRef` to customize the internal `Picker` to expose its functionality to the parent component, in this case, the `RichText`. I have utilized git's apply patch functionality to utilize these changes 💯